core/src/main/java/hudson/tasks/LogRotator.java

```diff
@@ -250,7 +250,7 @@ private boolean shouldKeepRun(Run r, Run lsb, Run lstb) {
             LOGGER.log(FINER, "{0} is not to be removed or purged of artifacts because it’s the last stable build", r);
             return true;
         }
-        if (r.isBuilding()) {
+        if (r.isLogUpdated()) {
```

Comment:
This is to avoid race conditions involving Pipeline builds. Previously it was possible for log rotation to delete builds which had not yet fully completed, which meant that a `build.xml` file could be written back out into the build directory _after_ log rotation had deleted the build. I will try to demonstrate this downstream in `workflow-job` once I have an incremental build here.

Reply:
Are you confident that this is safe for _freestyle_ builds? I.e., that `!logUpdated → !building` or conversely that `building → logUpdated`? The lifecycles for `AbstractBuild` and `WorkflowRun` are pretty different and the Javadoc has always been vague.

Reply:
I am pretty sure it is safe, but if you prefer we can switch to `isBuilding() || isLogUpdated()` like the check for the UI in Jelly (or maybe we should simplify that to only check `isLogUpdated` as well). See [here](https://github.com/jenkinsci/jenkins/blob/208238192f93e532284e4f8336e6a7849c50441b/core/src/main/java/hudson/model/Run.java#L518-L539) and the state enum [here](https://github.com/jenkinsci/jenkins/blob/208238192f93e532284e4f8336e6a7849c50441b/core/src/main/java/hudson/model/Run.java#L255-L278).

For `!logUpdated → !building`,`!isLogUpdated` is only true in `State.COMPLETED`, which ensures `!isBuilding` since `COMPLETED` is after `POST_PRODUCTION`.

The converse `building → logUpdated` should be ok as well, since if `isBuilding` then we must be in `State.CREATED` or `State.BUILIDING`, both of which are not `COMPLETED`, so `isLogUpdated` will be true.

For Pipeline things are simpler since `WorkflowRun.isLogUpdated` delegates to `WorkflowRun.isBuilding`.

Also just as a data point, if `isLogUpdated` could not be used on its own for non-pipelines, I think we would have seen it cause issues in tests due to the usage in [JenkinsRule.waitForCompletion](https://github.com/jenkinsci/jenkins-test-harness/blob/1542e2780eb84ab04b423a16b69da15ebea1e053/src/main/java/org/jvnet/hudson/test/JenkinsRule.java#L1633).

Reply:
(And as far as I know the states for a `Run` can only move forward, although looking at the code I guess a subclass that calls some of the protected methods incorrectly could cause random state changes.)

Reply:
> as far as I know the states for a `Run` can only move forward

That looks right.

> like the check for the UI in Jelly

Not sure what you mean. https://github.com/jenkinsci/jenkins/blob/208238192f93e532284e4f8336e6a7849c50441b/core/src/main/resources/hudson/model/Run/delete.jelly#L30 is only checking `building` and ought to be switch to only check `logUpdated`.

Reply:
🤦 I misread `keepLog` as `logUpdated`. I'll update that.

Reply:
I fixed that in [f676be5](https://github.com/jenkinsci/jenkins/pull/9810/commits/f676be52239c943c2b43e9869ca923e63658d9b0).